dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -0,0 +1,265 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Executor/PipelineExecutorContext.h>
+#include <Flash/Executor/toRU.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/IOPriorityQueue.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/MultiLevelFeedbackQueue.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h>
+#include <Flash/Pipeline/Schedule/Tasks/TaskHelper.h>
+
+namespace DB
+{
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submit(TaskPtr && task)
+{
+    std::lock_guard lock(mu);
+    submitWithoutLock(std::move(task));
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submit(std::vector<TaskPtr> & tasks)
+{
+    std::lock_guard lock(mu);
+    for (auto & task : tasks)
+    {
+        submitWithoutLock(std::move(task));
+    }
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submitWithoutLock(TaskPtr && task)
+{
+    if unlikely (is_finished)
+    {
+        FINALIZE_TASK(task);
+        return;
+    }
+    const auto & query_id = task->getQueryId();
+    if unlikely (cancel_query_id_cache.contains(query_id))
+    {
+        cancel_task_queue.push_back(std::move(task));
+        return;
+    }
+
+    // name can be empty, it means resource control is disabled.
+    const std::string & name = task->getResourceGroupName();
+
+    auto iter = resource_group_task_queues.find(name);
+    if (iter == resource_group_task_queues.end())
+    {
+        auto task_queue = std::make_shared<NestedQueueType>();
+
+        task_queue->submit(std::move(task));
+        resource_group_infos.push({LocalAdmissionController::global_instance->getPriority(name), task_queue, name});
+        resource_group_task_queues.insert({name, task_queue});
+    }
+    else
+    {
+        iter->second->submit(std::move(task));
+    }
+    cv.notify_one();
+}
+
+template <typename NestedQueueType>
+bool ResourceControlQueue<NestedQueueType>::take(TaskPtr & task)
+{
+    assert(task == nullptr);
+    std::unique_lock lock(mu);
+    while (true)
+    {
+        if (popTask(cancel_task_queue, task))
```

Comment:
Should check `finished` before?

Reply:
Maybe too many `is_finished` check? We already have two:
1. In `while (!resource_groups_info.empty())`
2. Before calling `cv.wait_for(lock, DEFAULT_WAIT_INTERVAL_WHEN_RUN_OUT_OF_RU)`

IMO, whether to check `is_finished` before `pop cancel task queue` has little impact.

Reply:
But not checking "finished" contradicts the semantics of "take".
https://github.com/pingcap/tiflash/blob/6f7c6472e1b59abb7e101f3b708094e81d52c568/dbms/src/Flash/Pipeline/Schedule/TaskQueues/TaskQueue.h#L43-L44

Reply:
Maybe checking is_finished in L81 is enough, like MLFQ
https://github.com/pingcap/tiflash/blob/6f7c6472e1b59abb7e101f3b708094e81d52c568/dbms/src/Flash/Pipeline/Schedule/TaskQueues/MultiLevelFeedbackQueue.cpp#L178-L180

Reply:
Since the mutex is used in finish, it doesn't make sense to check is_finished in the second while.
There is condition_cv.wait in the first while, which can give up the control of mutex

Reply:
ok